File: android/feature/photos-ui/src/main/kotlin/xyz/klinik/feature/photos/ui/PhotoScreens.kt
package xyz.klinik.feature.photos.ui

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectHorizontalDragGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Slider
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clipToBounds
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import xyz.klinik.design.Tokens
import xyz.klinik.design.klinikColor
import xyz.klinik.feature.photos.ComparisonPair
import xyz.klinik.feature.photos.GalleryPhase
import xyz.klinik.feature.photos.GalleryState
import xyz.klinik.network.ClinicalPhoto
import xyz.klinik.network.PhotoCategory
import xyz.klinik.network.UiText

/** Text the screen needs, resolved by the caller from string resources. */
data class PhotoStrings(
    val empty: String,
    val notFound: String,
    val retry: String,
    val compare: String,
    val compareSlider: String,
    val before: String,
    val after: String,
    val consentGiven: String,
    val clinicalUseOnly: String,
    val noBodyArea: String,
    val categoryName: (PhotoCategory) -> String,
    val message: (String) -> String,
)

@Composable
fun PhotoGalleryScreen(
    state: GalleryState,
    strings: PhotoStrings,
    /**
     * Decoded image for a photo, supplied by the caller.
     *
     * Kept out of here so the screen carries no image-loading dependency and
     * never holds a signed URL longer than it draws it — those URLs are short
     * lived on purpose.
     */
    imageFor: (String) -> ImageBitmap?,
    onRetry: () -> Unit,
    onSelectArea: (String) -> Unit,
    onCompare: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Surface(color = klinikColor("background"), modifier = modifier.fillMaxSize()) {
        when (val phase = state.phase) {
            GalleryPhase.Loading -> Centered { CircularProgressIndicator() }

            GalleryPhase.Empty -> Centered {
                Text(strings.empty, color = klinikColor("textSecondary"))
            }

            GalleryPhase.NotFound -> Centered {
                Text(strings.notFound, color = klinikColor("textSecondary"))
            }

            is GalleryPhase.Failed -> Centered {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
                ) {
                    Text(strings.message(phase.messageKey), color = klinikColor("critical"))
                    TextButton(onClick = onRetry) { Text(strings.retry) }
                }
            }

            GalleryPhase.Loaded ->
                Gallery(state, strings, imageFor, onSelectArea, onCompare)
        }
    }
}

@Composable
private fun Gallery(
    state: GalleryState,
    strings: PhotoStrings,
    imageFor: (String) -> ImageBitmap?,
    onSelectArea: (String) -> Unit,
    onCompare: () -> Unit,
) {
    Column(modifier = Modifier.fillMaxSize()) {
        state.error?.let { error ->
            val text = when (error) {
                is UiText.Key -> strings.message(error.key)
                is UiText.Literal -> error.text
            }

            Text(
                text,
                color = klinikColor("critical"),
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(Tokens.Spacing.lg)
                    .semantics { contentDescription = text },
            )
        }

        Row(
            modifier = Modifier.padding(horizontal = Tokens.Spacing.lg),
            horizontalArrangement = Arrangement.spacedBy(Tokens.Spacing.sm),
        ) {
            state.groups.forEach { group ->
                val label = group.bodyArea ?: strings.noBodyArea

                if (group.id == state.selectedGroup?.id) {
                    Button(
                        onClick = { onSelectArea(group.id) },
                        modifier = Modifier.height(Tokens.minimumTouchTarget),
                    ) {
                        Text(label)
                    }
                } else {
                    TextButton(
                        onClick = { onSelectArea(group.id) },
                        modifier = Modifier.height(Tokens.minimumTouchTarget),
                    ) {
                        Text(label)
                    }
                }
            }
        }

        LazyColumn(modifier = Modifier.weight(1f)) {
            items(state.selectedGroup?.photos.orEmpty(), key = { it.id }) { photo ->
                PhotoRow(photo, strings, imageFor(photo.id))
            }
        }

        // Offered only when there are two photos to compare.
        if (state.comparison != null) {
            Button(
                onClick = onCompare,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(Tokens.Spacing.lg)
                    .height(Tokens.minimumTouchTarget),
            ) {
                Text(strings.compare)
            }
        }
    }
}

@Composable
private fun PhotoRow(photo: ClinicalPhoto, strings: PhotoStrings, image: ImageBitmap?) {
    val consentText = if (photo.hasUsageConsent) strings.consentGiven else strings.clinicalUseOnly
    val label = photo.phaseLabel ?: strings.categoryName(photo.category)

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(Tokens.Spacing.lg)
            .semantics(mergeDescendants = true) {
                contentDescription = "$label, ${photo.takenAt}, $consentText"
            },
        verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs),
    ) {
        PhotoImage(image, Modifier.fillMaxWidth().height(200.dp))

        Text(label, color = klinikColor("textPrimary"), modifier = Modifier.clearAndSetSemantics {})

        // Said on the photo itself rather than in a settings screen: whether
        // this image may be used outside the clinic is a fact about the image.
        Text(
            consentText,
            color = klinikColor(if (photo.hasUsageConsent) "success" else "textSecondary"),
            modifier = Modifier.clearAndSetSemantics {},
        )
    }
}

/** Two photos with a divider the reader drags (spec M7: slider comparison). */
@Composable
fun PhotoComparisonScreen(
    pair: ComparisonPair,
    strings: PhotoStrings,
    imageFor: (String) -> ImageBitmap?,
    modifier: Modifier = Modifier,
) {
    var split by remember { mutableFloatStateOf(0.5f) }

    Surface(color = klinikColor("background"), modifier = modifier.fillMaxSize()) {
        Column(
            modifier = Modifier.padding(Tokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
        ) {
            BoxWithConstraints(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(360.dp)
                    .clipToBounds()
                    .pointerInput(Unit) {
                        detectHorizontalDragGestures { change, _ ->
                            split = (change.position.x / size.width).coerceIn(0f, 1f)
                        }
                    }
                    // The images are decoration here; the labels below name
                    // which is which, and the slider carries the interaction.
                    .clearAndSetSemantics {},
            ) {
                val full = maxWidth

                PhotoImage(imageFor(pair.after.id), Modifier.fillMaxSize())

                // The "before" image is drawn at full width inside a narrower
                // box that clips it. Scaling it to the box instead would squash
                // the picture as the divider moved, which is exactly the
                // distortion a before/after comparison must not introduce.
                Box(modifier = Modifier.width(full * split).fillMaxHeight().clipToBounds()) {
                    PhotoImage(imageFor(pair.before.id), Modifier.width(full).fillMaxHeight())
                }

                Box(
                    modifier = Modifier
                        .offset(x = full * split)
                        .width(2.dp)
                        .fillMaxHeight()
                        .background(klinikColor("accent")),
                )
            }

            // The divider is a drag gesture, which a screen reader cannot
            // perform. The same comparison is reachable as a value it can set.
            Slider(
                value = split,
                onValueChange = { split = it },
                modifier = Modifier.semantics {
                    contentDescription = strings.compareSlider
                },
            )

            Row(modifier = Modifier.fillMaxWidth()) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(strings.before, color = klinikColor("textSecondary"))
                    Text(
                        pair.before.phaseLabel ?: strings.categoryName(pair.before.category),
                        color = klinikColor("textPrimary"),
                    )
                }

                Column(horizontalAlignment = Alignment.End) {
                    Text(strings.after, color = klinikColor("textSecondary"))
                    Text(
                        pair.after.phaseLabel ?: strings.categoryName(pair.after.category),
                        color = klinikColor("textPrimary"),
                    )
                }
            }
        }
    }
}

@Composable
private fun PhotoImage(image: ImageBitmap?, modifier: Modifier) {
    if (image == null) {
        // A grey rectangle rather than an icon: the caller may still be minting
        // the signed URL, and a failure symbol for a photo that is merely late
        // reads as a photo that is gone.
        Box(modifier = modifier.background(klinikColor("surface")))
        return
    }

    Image(
        bitmap = image,
        contentDescription = null,
        contentScale = ContentScale.Crop,
        modifier = modifier,
    )
}

@Composable
private fun Centered(content: @Composable () -> Unit) {
    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { content() }
}

File: ios/Sources/KlinikPhotosFeature/PhotoScreens.swift
import SwiftUI
import KlinikAPI
import KlinikCore
import KlinikDesign

/// The before/after gallery, and the comparison it opens into (spec M7).
public struct PhotoGalleryView: View {
    @Environment(\.colorScheme) private var scheme

    private let model: PhotoGalleryModel
    private let linkFor: @Sendable (String) async -> URL?

    @State private var state = GalleryState()
    @State private var comparing = false

    /// - Parameter linkFor: mints a short-lived signed URL for a photo. Supplied
    ///   by the caller so the screen never holds a URL longer than it draws it.
    public init(
        model: PhotoGalleryModel,
        linkFor: @escaping @Sendable (String) async -> URL?
    ) {
        self.model = model
        self.linkFor = linkFor
    }

    public var body: some View {
        VStack(spacing: 0) {
            content
        }
        .background(Tokens.Palette.background.resolve(for: scheme))
        .task { await refresh { await model.load() } }
        .sheet(isPresented: $comparing) {
            if let pair = state.comparison {
                PhotoComparisonView(pair: pair, linkFor: linkFor)
            }
        }
    }

    @ViewBuilder
    private var content: some View {
        switch state.phase {
        case .loading:
            Spacer()
            ProgressView().accessibilityLabel(L10n.string("common.loading"))
            Spacer()

        case .empty:
            MessageState(icon: "photo.on.rectangle", text: L10n.string("photo.empty"))

        case .notFound:
            MessageState(icon: "questionmark.folder", text: L10n.string("error.notFound"))

        case .failed(let message):
            MessageState(
                icon: Tokens.State.labCritical.iconName,
                text: message,
                retryTitle: L10n.string("common.retry")
            ) {
                await refresh { await model.load() }
            }

        case .loaded:
            gallery
        }
    }

    private var gallery: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                if let error = state.error {
                    ErrorBanner(message: error)
                }

                Picker("", selection: areaSelection) {
                    ForEach(state.groups) { group in
                        Text(group.bodyArea ?? L10n.string("photo.noBodyArea")).tag(group.id)
                    }
                }
                .pickerStyle(.menu)
                .accessibilityLabel(L10n.string("photo.bodyArea"))

                if let group = state.selectedGroup {
                    ForEach(group.photos) { photo in
                        PhotoRow(photo: photo, linkFor: linkFor)
                    }
                }

                // Offered only when there are two photos to compare: a slider
                // over a single image does nothing and suggests there is a
                // change to see.
                if state.comparison != nil {
                    PrimaryButton(
                        title: L10n.string("photo.compare"),
                        isBusy: false,
                        isEnabled: true
                    ) {
                        comparing = true
                    }
                }
            }
            .padding(Tokens.Spacing.lg)
        }
    }

    private var areaSelection: Binding<String> {
        Binding(
            get: { state.selectedArea ?? state.groups.first?.id ?? "" },
            set: { area in Task { await refresh { await model.select(area: area) } } }
        )
    }

    private func refresh(_ work: () async -> Void) async {
        await work()
        state = await model.currentState()
    }
}

struct PhotoRow: View {
    @Environment(\.colorScheme) private var scheme

    let photo: ClinicalPhoto
    let linkFor: @Sendable (String) async -> URL?

    @State private var url: URL?

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.xs) {
            SignedImage(url: url)
                .frame(height: 200)
                .clipShape(RoundedRectangle(cornerRadius: Tokens.Radius.md))

            HStack(spacing: Tokens.Spacing.sm) {
                Text(photo.phaseLabel ?? photo.category.localizedName)
                    .font(Tokens.Typography.subheadingRelative)
                    .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

                Spacer()

                Text(photo.takenAt, style: .date)
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
            }

            // Said on the photo itself rather than buried in a settings screen:
            // whether this image may be used outside the clinic is a fact about
            // the image, and the person looking at it is the one who might.
            Label(
                L10n.string(photo.hasUsageConsent ? "photo.consentGiven" : "photo.clinicalUseOnly"),
                systemImage: photo.hasUsageConsent ? "checkmark.shield" : "lock.shield"
            )
            .font(Tokens.Typography.captionRelative)
            .foregroundStyle(
                (photo.hasUsageConsent ? Tokens.Palette.success : Tokens.Palette.textSecondary)
                    .resolve(for: scheme)
            )
        }
        .task { url = await linkFor(photo.id) }
        .accessibilityElement(children: .combine)
    }
}

/// Two photos with a divider the reader drags (spec M7: slider comparison).
public struct PhotoComparisonView: View {
    @Environment(\.colorScheme) private var scheme

    private let pair: ComparisonPair
    private let linkFor: @Sendable (String) async -> URL?

    @State private var beforeURL: URL?
    @State private var afterURL: URL?
    @State private var split: CGFloat = 0.5

    public init(pair: ComparisonPair, linkFor: @escaping @Sendable (String) async -> URL?) {
        self.pair = pair
        self.linkFor = linkFor
    }

    public var body: some View {
        VStack(spacing: Tokens.Spacing.lg) {
            GeometryReader { geometry in
                ZStack(alignment: .leading) {
                    SignedImage(url: afterURL)

                    SignedImage(url: beforeURL)
                        .frame(width: geometry.size.width * split, alignment: .leading)
                        .clipped()

                    Rectangle()
                        .fill(Tokens.Palette.accent.resolve(for: scheme))
                        .frame(width: 2)
                        .offset(x: geometry.size.width * split - 1)
                }
                .contentShape(Rectangle())
                .gesture(
                    DragGesture()
                        .onChanged { value in
                            split = min(max(value.location.x / geometry.size.width, 0), 1)
                        }
                )
            }
            .frame(height: 360)
            .clipShape(RoundedRectangle(cornerRadius: Tokens.Radius.md))

            // The slider is a drag gesture, which VoiceOver cannot perform. The
            // same comparison is reachable as a value a rotor can adjust.
            Slider(value: $split, in: 0...1)
                .accessibilityLabel(L10n.string("photo.compareSlider"))
                .accessibilityValue("\(Int(split * 100))%")

            HStack {
                label(pair.before, key: "photo.before")
                Spacer()
                label(pair.after, key: "photo.after")
            }
        }
        .padding(Tokens.Spacing.lg)
        .background(Tokens.Palette.background.resolve(for: scheme))
        .task {
            beforeURL = await linkFor(pair.before.id)
            afterURL = await linkFor(pair.after.id)
        }
    }

    private func label(_ photo: ClinicalPhoto, key: String) -> some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.xxs) {
            Text(L10n.string(key))
                .font(Tokens.Typography.captionRelative)
                .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))

            Text(photo.phaseLabel ?? photo.category.localizedName)
                .font(Tokens.Typography.calloutRelative)
                .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))
        }
        .accessibilityElement(children: .combine)
    }
}

/// An image behind a short-lived signed URL, with a placeholder while it loads.
struct SignedImage: View {
    @Environment(\.colorScheme) private var scheme

    let url: URL?

    var body: some View {
        Group {
            if let url {
                AsyncImage(url: url) { phase in
                    switch phase {
                    case .success(let image):
                        image.resizable().scaledToFill()
                    case .failure:
                        placeholder(icon: "exclamationmark.triangle")
                    default:
                        placeholder(icon: "photo")
                    }
                }
            } else {
                placeholder(icon: "photo")
            }
        }
        .frame(maxWidth: .infinity)
        .clipped()
    }

    private func placeholder(icon: String) -> some View {
        ZStack {
            Tokens.Palette.surface.resolve(for: scheme)
            Image(systemName: icon)
                .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                // A stand-in for a photograph that has not loaded. The photo's
                // own row carries the description; this is the empty frame.
                .accessibilityHidden(true)
        }
    }
}
